ambry-router/src/test/java/com.github.ambry.router/GetOperationTest.java

```diff
@@ -0,0 +1,474 @@
+   */
+  @Test
+  public void testSuccessInThePresenceOfVariousErrors()
+      throws Exception {
+    GetBlobInfoOperation op =
+        new GetBlobInfoOperation(routerConfig, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            operationCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    ArrayList<MockServer> mockServers = new ArrayList<>(mockServerLayout.getMockServers());
+    // set the status to various server level or partition level errors (not Blob_Deleted or Blob_Expired).
+    mockServers.get(0).setServerErrorForAllRequests(ServerErrorCode.Unknown_Error);
+    mockServers.get(1).setServerErrorForAllRequests(ServerErrorCode.Unknown_Error);
+    mockServers.get(2).setServerErrorForAllRequests(ServerErrorCode.IO_Error);
+    mockServers.get(3).setServerErrorForAllRequests(ServerErrorCode.No_Error);
+    mockServers.get(4).setServerErrorForAllRequests(ServerErrorCode.Data_Corrupt);
+    mockServers.get(5).setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+    mockServers.get(6).setServerErrorForAllRequests(ServerErrorCode.No_Error);
+    mockServers.get(7).setServerErrorForAllRequests(ServerErrorCode.Disk_Unavailable);
+    mockServers.get(8).setServerErrorForAllRequests(ServerErrorCode.Unknown_Error);
+
+    // set the status of one of the servers in the local datacenter to success (we depend on an actual put of the blob
+    // in order to get back the BlobInfo, and the put would have gone only to the local datacenter).
+    for (int i = 0; i < mockServers.size(); i++) {
+      MockServer mockServer = mockServers.get(i);
+      if (mockServer.getDataCenter().equals(routerConfig.routerDatacenterName)) {
+        mockServer.setServerErrorForAllRequests(ServerErrorCode.No_Error);
+        break;
+      }
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    assertSuccess();
+  }
+
+  /**
+   * Submit all the requests that were handed over by the operation and wait until a response is received for every
+   * one of them.
+   * @param requestList the list containing the requests handed over by the operation.
+   * @return the list of responses from the network client.
+   * @throws IOException
+   */
+  private List<ResponseInfo> sendAndWaitForResponses(List<RequestInfo> requestList)
+      throws IOException {
+    List<ResponseInfo> responseList = new ArrayList<>();
+    int sendCount = requestList.size();
+    responseList.addAll(networkClient.sendAndPoll(requestList));
+    requestList.clear();
```

Comment:
the contents of `requestList` is handed over to the `networkClient` in the first `sendAndPoll()`. We then clear it so that we don't hand them over again. Subsequent calls to `sendAndPoll` are just to poll for responses - not to send anything.